### PR TITLE
Maintenance/technical refactoring

### PR DESCRIPTION
**Pull request by JonasJurczok**
@Tarrke @hoylemd I'm looking for your feedback here :)
Please write what you think and if it actually helps in understanding the mod

Goals:
* It should be easier to find the implementation of certain features
* Making changes to features should be touching less files/places
* UI strictly separated from each other and other code
* A lot of cleanup in naming etc.

Changes:
* All business logic is in the /features folder
* All UI related logic is in the /ui folder
* All function names follow the todo.*feature*_*action* pattern
* All element names follow the todo_*feature*_*meaning* pattern
* Tasks are now called tasks everywhere (item is gone)
* Frames are static, always visible frames (the main UI)
* Dialogs are things that pop up (add, edit, import, ...)
* It is now possible to have one edit and one add dialog open at the same time (side effect of renaming everything)
* The main todo file is now only an anchor/event handler. Everything else is done elsewhere
* Files now contain comments what is in them -.-

Downsides:
* Lots of files
* Very distributed code base

### CODE REVIEW

**Comment by JonasJurczok** [comment]
After looking at the changes here it is probably best to checkout the branch and then play around with it a little :)

**Comment by hoylemd** [comment]
@Tarrke I'll try to make time to review it in the next week or so.  I've got a conference this weekend though, so depending on how interesting the talks are, I'll either do it then or have no time at all :p

**Comment by Tarrke** [comment]
Hi guys.

I have an rpg session this weekend and need to prep my char. Have a nice
conference on your side.

However I'll review this tomorrow and Friday from work. I get the main
point here and I'm full of hope for this PR. At this point I only have one
thing to add. If you really want to reduce the number of files implied in a
change, just throw everything into a single file.

Le mer. 7 nov. 2018 à 18:47, Michael D. Hoyle <notifications@github.com> a
écrit :

**Comment by JonasJurczok** [comment]
I mainly wanted to reduce the line count per file. And separate things where they belong.
So.. yes there are much more files now, but I think its worth it.

**Comment by Tarrke** [comment]
You know my last sentence was a joke right?

**Comment by JonasJurczok** [comment]
> You know my last sentence was a joke right?

no m( :D

File: src/todo/features/add_task.lua [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    todo.log("Creating task by player " .. player.name)
+
+    local dialog = todo.get_add_dialog(player)
+    if (dialog == nil) then
+        return
+    end
+
+    local task_data, should_add_to_top = todo.get_task_from_add_dialog(dialog)
+    local task = todo.assemble_task(task_data, player)
+
```
Comment (Tarrke) [inline src/todo/features/add_task.lua]: This line is too much.

File: src/todo/features/add_task.lua [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    local assignees = dialog.todo_add_task_table["todo_add_assignee_drop_down"]
+    local assignee
+    if (assignees.selected_index > 1) then
+        assignee = assignees.items[assignees.selected_index]
+    end
+
+    local add_top_checkbox = dialog.todo_add_task_table["todo_add_top"]
+    local should_add_to_top = add_top_checkbox.state
+
+    local task = { ["title"] = title, ["task"] = text, ["assignee"] = assignee}
```
Comment (Tarrke) [inline src/todo/features/add_task.lua]:
```suggestion
    local task = { ["title"] = title, ["task"] = text, ["assignee"] = assignee }
```

File: src/todo/logging.lua [hunk 3]
```diff
@@ -0,0 +1,21 @@
+-- TODO: String format support missing
```
Comment (Tarrke) [inline src/todo/logging.lua]: This should be a new issue.

File: spec/29_import_export_feature.lua [hunk 4]
```diff
@@ -1,12 +1,12 @@
 feature("#29 import and export tasks", function()
 
     before_scenario(function()
-        when(todo, "show_button"):then_return(true)
-        todo.maximize(game.players[1])
+        when(todo, "is_show_maximize_button"):then_return(true)
```
[the remark this answers is not in the document]
Comment (hoylemd) [reply spec/29_import_export_feature.lua]: 'is_maximize_button_shown' would be best of those options, but  'should_show_maximize_button' is another option.

**Comment by Tarrke** [comment]
When do you plan to push this to master? I'm planning to add some feature once this goes live.

File: spec/60_show_created_by_feature.lua [hunk 5]
```diff
@@ -34,25 +35,27 @@ feature("#60 Show created/updated by", function()
...
+        local task = todo.assemble_task(task_template, player1)
+        todo.save_task_to_open_list(task)
         todo.refresh_task_table(player)
 
-        todo.create_add_edit_frame(player, task)
-        local frame = todo.get_add_edit_frame(player)
-        local name_label = frame["todo_add_task_table"]["todo_created_by_playername"]
+        faketorio.click("todo_open_edit_dialog_button_" .. task.id, player)
+        local dialog = todo.get_edit_dialog(player)
+        local name_label = dialog["todo_edit_task_table"]["todo_edit_created_by_playername"]
```
[the remark this answers is not in the document]
Comment (JonasJurczok) [reply spec/60_show_created_by_feature.lua]: I really just did this for readability. And its in a test so... :)

File: spec/60_show_created_by_feature.lua [hunk 6]
```diff
@@ -34,25 +35,27 @@ feature("#60 Show created/updated by", function()
...
-        todo.update(name_label, task.id, player2)
-        todo.create_add_edit_frame(player, task)
-        frame = todo.get_add_edit_frame(player)
-        assert(frame["todo_add_task_table"]["todo_updated_by_playername"].caption == player2.name)
-    end)
+        todo.edit_persist_task_changes(player, task.id, player2)
+        todo.create_add_task_dialog(player, task)
+        dialog = todo.get_add_dialog(player)
+        assert(dialog["todo_edit_task_table"]["todo_edit_updated_by_playername"].caption == player2.name)
+    end)]]--
```
[the remark this answers is not in the document]
Comment (JonasJurczok) [reply spec/60_show_created_by_feature.lua]: Because i'm blind :D

File: spec/UI_feature.lua [hunk 7]
```diff
@@ -1,22 +1,24 @@
 local task_num = 1
-function add_task(text)
+function add_task(player, text)
```
[the remark this answers is not in the document]
Comment (JonasJurczok) [reply spec/UI_feature.lua]: Yes. I also thought about that.. but for the moment I don't want to open up that box of pandorra :D

File: src/todo/features/add_task.lua (hunk 1 above)
Comment (JonasJurczok) [reply src/todo/features/add_task.lua]: What do you mean?

File: src/todo/logging.lua (hunk 3 above)
Comment (JonasJurczok) [reply src/todo/logging.lua]: Yeah. I just noticed this but yes.. the logging will get some love at some point but not here :)

**Comment by hoylemd** [comment]
Oh no! I was too slow.  I have about half of a review finished, but it's pretty much entirely naming and code organization stuff - functionally it looks good to me.  Would you like me to finish the review, and you can use that feedback next time you do a refactor?

**Pull request merged**